Bug fixes

andy013 opened the pull request:
Hey, I had a go at fixing #75. It seems to be working but you better triple check the code though because the part I deleted I'm not 100% sure what it did. (this was change to common.js)

I also fixed an issue I discovered where if you set a restricted number of players the game would not start even when you reached your minimum. To repo just start a game with 6 players min and a max >6. Join the game with 6 players and it will just sit idle and do nothing. Yet, if one player leaves and then rejoins the game will begin. I also tried to update the player counter so that it accurately reflected the number of players required for the game to begin. (this was change to user-events.js)

andy013 commented:
Ah, I just realized what the code did. I never though of needing to cancel the start of the game but if you have only 5-7 players selected then when a 6th joins you need to stop the count down. Let me review this and see if I can fix it.

cozuya commented:
Yep reference (#81).  I'm super cautious about back end code like this - I would love to make it bullet proof, but theres a lot of situations that needs to be addressed before any back end changes.  I will def look at your PR though.

andy013 commented:
I'll try and update this and resubmit if I figure it out.

andy013 commented:
I spent a little time on this and I think its working well now. As far as I can tell the parts I changed to fix #75 were treating special election presidents differently and allowing them to choose anyone as their chancellor. It's only the person who has the presidential power who can select anyone but once he has chosen someone the normal rules apply for them.

For the second bug I tried to rewrite some of the code so that it would better handle players leaving before the game begins. Now, if they leave before the boards are flipped then they won't show up as disconnected and stay in their seat. Instead they will just be removed and the normal rules for whether the game should start or not will be followed (hopefully). I also tried to update the player counter so that when players leave and join it will update accordingly rather than just saying "Waiting for more players...". The only issue I ran into was that I wasn't sure how to force the game to refresh. I figured out that if you type something in chat then it will update. At lines 1022 and 1027 I copied a bit of code that seems to do this automatically but since I'm not confident with my understanding of it, you may want to give it a closer look. Those lines were dealing with when players leave, but there is a different function handling disconnects. I tried using the same piece of code but couldn't get it to refresh upon a player disconnecting and had to resort to posting a message in the chat. It still works fine but it would be nice if it automatically updated when a player left.

cozuya commented:
Awesome does it cover the cases noted in #137?  I will take a look this week.

andy013 commented:
Yeah, I think it fixes those issues. Although I guess it's possible a player could leave at just the right time and it may still keep counting, but I didn't see it in testing so far.

cozuya commented:
Thats great its a nasty and complicated one to track down that I haven't had time for.  Re: the gameupdates thing yeah I noticed that too in that in some situations a gameupdate broadcast to room is needed to right the UI.  I don't think its worth a ton of time investing though as they are 1-3 per second it feels like in real games.

andy013 commented:
I think I managed to fix the chat sorting issue. The video Dennis posted really helped as I then figured out how to repo it. It was a bit of a pain in the ass as just comparing the 2 chats wasn't working because some of them were strings and others were arrays of objects. I ended up using .reduce to turn the objects into one long string and then comparing them that way. I don't know if there's and easier way to do it.

cozuya commented:
Very cool andy.  I want to finish work on my routing issue and then I'll review this.  Might be a few days.

andy013 commented:
No worries man :+1:

andy013 commented:
I think I found the cause of #277. It wasn't to do with any of my changes. It seems like what is happening is that the old session is being saved after the players status is changed to 'connected: false'. Then when they reconnect, this is overwriting their status and sticking that way. If you log out cleanly then there is no old session and so it's not a problem. It's only an issue if you log in on another window without logging out. 

I commented out the line that seems to be responsible and it seems to be working, but I don't really know what the purpose of saving the old session was, so it might have unintended consequences.

cozuya commented:
Awesome that you found a fix to that veto issue.  Since that it is pretty critical, I'd like to review this PR and patch soon.  Is this ready for review Andy?  If so, you should pull master and resolve merge conflicts.  Thanks!

andy013 commented:
I managed to fix the president not being able to vote on the veto.

andy013 commented:
I noticed that you added some code in user-events to the end of `handleSocketDisconnect`. It looks like you moved the call to `sendUserList()` inside an if statement. Specifically it's now inside `if (passport && Object.keys(passport).length)`. Just want to check as their is an extra curly brace in my version now and I'm not sure whether to remove it or not.

cozuya commented:
It'll probably crash if its extra :D I haven't committed anything to this branch if thats what you mean - I'd have to look at the log/blame to see if that functions been worked on lately but that one is usually pretty solid.

andy013 commented:
I figured it out. For some reason git was including one of the closing braces inside my part of the conflict, even though both files shared it. So when I deleted my code it also removed the brace. It should be good to go now.

cozuya commented:
Great!  and thanks for the summary I was about to suggest that.  Can I impose on you one more time - can you add the github issue number next to the above update (if they exist)?  Thanks!  I'll look to review over the next couple days.

cozuya commented:
Looks great Andy!  Thanks again.  2 things: would you mind following the contributing guidelines? https://github.com/cozuya/secret-hitler/blob/master/CONTRIBUTING.md especially prettier will keep our code looking the same.

As far as the user-events socket disconnection thing, that emit does this and then this:

https://github.com/cozuya/secret-hitler/blob/master/src/frontend-scripts/components/App.jsx#L79

https://github.com/cozuya/secret-hitler/blob/master/routes/index.js#L96

obviously making it so that only one browser window can be on /game per account.  That shouldn't, itself, cause it to appear disconnected, it should be something else.  I think.  When I try while that line is commented out I still can't make a connected account appear grey in chat.

Going to patch in a couple hours.

andy013 commented:
Hey, I did use prettier but I tried to be smart and installed a plug-in for sublime so I wouldn't need to use the CLI. I think sublime has a different default tab-width setting so it screwed with the formatting even though all the settings were the same. I also forgot to run it after resolving the conflicts and I just edited them through chrome so I might have messed it up manually that way. Sorry about that. 

Maybe I'll check out that disconnect bug later. When I was testing I noticed that the players status was being set to connected: false at some point and commenting that line out seemed to resolve it. I can't seem to reproduce appearing grey in chat without it.

cozuya commented:
No big deal re: prettier, obviously I just hit save and it works just fine.  I recently switched from sublime to VS code and while there's pluses and minuses, at least VS code is under very active development (and MIT).  I have a repo on github for settings if you're interested.

Disconnect code is def not perfect =/ and tough to figure out and track down, especially without a dedicated test/stage environment.

Anything you do to help is great!  Thanks again.

andy013 commented:
I noticed an issue with the countdown going below 4 and the game not starting. I'm not sure how that can happen. The only thing I can think of is that multiple players are joining at the same time and it is calling the startCountdown function more than once. I guess I never saw it in testing because I can only join games one at a time. Maybe you could add this to the top of startCountdown `	if (game.gameState.isStarted === true) {
		return;
	}` I don't know if it'll fix it or not, just a quick guess for now.
